content/post/pga_history.md

```diff
@@ -0,0 +1,91 @@
+Check out the [announcement post](https://blog.sourced.tech/post/announcing-pga/) for more information.
+You should check out [Engine](https://github.com/src-d/engine) which allows
+to run SQL queries on top the PGA and do other cool things.
+
+## Origins
+
+We've wanted to release all the Git repositories we are able to clone since the source{d}'s beginning. Back in 2016 we stored millions of Git [packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) in Google Cloud Storage and the metadata in MongoDB. It worked well for us but that scheme did not allow to share the data. Then we decided to go baremetal in 2017 and started to build our infrastructure from scratch. One of the
+important changes was the creation of [Siva](https://github.com/src-d/go-siva) Git repository archive format.
+Siva resembles [Tar](https://www.gnu.org/software/tar/) but the index is placed at the end of the file to allow
+cheap append operation. We re-invented the concept of a "rooted repository" where all branches from the same root are placed together, and therefore the corresponding packfiles are appended to the same Siva file while we are massively cloning Git repositories. It is not unique - GitHub has a similar approach to storing forks
+internally. The main benefit is re-using the Git objects between branches and thus reducing the size on disk.
+
+So when our next-gen retrieval pipeline emerged, we realized that the dataset time has come. It is surely
+difficult to open access to the whole Git world (we clone from BitBucket, Savannah, servers in the wild, etc.)
+because the size is too big, so we started from putting a threshold to the number of stargazers on GitHub
+and picking the most popular projects. This threshold was set to **50** from Vadim's
+historical default limit in [GitHubStars](https://github.com/vmarkovtsev/githubstars).
+The plan was to distribute the Siva files generated from those popular repositories.
+Later we realized that having only the Siva files is not enough - people wanted to tighten the stars limit,
+filter by main programming language or by license. The idea of an index file appeared where we would place
+various metadata, either existing or mined after cloning.
+
+Cloning Git repositories is ~hard~ 🔥HELL🔥:
+
+* Integration between Gerrit and GitHub produces hundreds of standalone Git references. E.g. [google/angle](https://github.com/google/angle) becomes spread over 6,600 Siva files, each with a tiny single ref coming from
+a Gerrit review.
+* [Git Bomb](https://github.com/Katee/git-bomb) - an "uncloneable" repository. By the way, we cloned it.
+* Some repositories have packfiles in the [old deprecated format]. go-git handles them.(https://github.com/git/git/blob/master/Documentation/technical/pack-format.txt#L39).
+* Load balancing is difficult. 90% of the repositories are compact and fly fast, however, the rest 10% occupy 50% of the dataset size and are damn slow to process.
```

Comment:
the remaining 10% take up 50% of the total size of the dataset and are really lengthy to process